### PR TITLE
remove rehash! from Dict constructor

### PR DESCRIPTION
**Pull request by bkamins**
This PR follows https://discourse.julialang.org/t/is-there-a-bug-in-dict/6672/4.
It implements:
* `Dict{K,V}(d::Dict{K,V})` constructor performs a bare copy of `d` (without `rehash!`);
* new test that currently fails and should pass after the PR;
* rename `ht_keyindex2` to `ht_keyindex2!` (this is not strictly necessary for this PR so can be omitted, but it cleans up the related code, as `ht_keyindex2` in some cases may call `rehash!` thus it may mutate the passed dictionary)

I recommend to remove `rehash!` from the constructor. Another approach would be to rehash a newly created `Dict`:

    function Dict{K,V}(d::Dict{K,V}) where V where K
        d2 = new(copy(d.slots), copy(d.keys), copy(d.vals), d.ndel, d.count, d.age,
                d.idxfloor, d.maxprobe)
        d2.ndel > 0 && rehash!(d2)
        d2
    end

but I do not see if there would be any huge benefit of this - `rehash!` will be called anyway if needed by `setindex!` or `get!` and it adds performance penalty during construction.

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -104,12 +104,8 @@ mutable struct Dict{K,V} <: Associative{K,V}
...
     end
     function Dict{K,V}(d::Dict{K,V}) where V where K
-        if d.ndel > 0
-            rehash!(d)
-        end
-        @assert d.ndel == 0
-        new(copy(d.slots), copy(d.keys), copy(d.vals), 0, d.count, d.age, d.idxfloor,
-            d.maxprobe)
+        new(copy(d.slots), copy(d.keys), copy(d.vals), d.ndel, d.count, d.age,
+            d.idxfloor, d.maxprobe)
```
Comment (stevengj) [inline base/dict.jl]: Should we do `dnew = new(...)` and then `rehash!(dnew)` if `d.ndel > 0`?

**Comment by stevengj** [comment]
Note that the `rehash!` was added by @JeffBezanson in 41f3cf8fd176b92e1b49d89a84e1b495c8889bf1 ... I tend to think we should `rehash!` the newly created dictionary if needed, but I agree that we shouldn't modify the old dict.

File: base/dict.jl (hunk 1 above)
Comment (bkamins) [reply base/dict.jl]: This is exactly what I have considered (as given in the comment to the PR). I have decided not to add it as it introduces a performance penalty for `Dict` creation. Probably @JeffBezanson can provide the guidance here.
Another approach would be to test `d.ndel > "some value greater than 0"` (i.e. if we have a lot of deleted entries we do `rehash!`, and if there are only a few we leave it without rehashing).

Comment (bkamins) [reply base/dict.jl]: As a second thought - my recommendation would be to remove `rehash!` from the constructor. Its main benefit is that later we do not have to `rehash!` both dictionaries twice. But this is something that we cannot avoid anyway.
If we believe that being able to a clean dictionary is an important feature I would rather recommend to:
* rename `rehash!` to `_rehash!`;
* expose a new `rehash!` as an exported function and define it for `Dict` and `Set`. This new `rehash!` should perform an additional check if the size is not decreased (as `_rehash!` is designed for growing only).

**Comment by JeffBezanson** [comment]
Yes, this is ok as a (quasi-)bug fix. As a later enhancement, if the fraction of deleted keys is high we should probably copy the pairs by `for (k,v) in d; newd[k]=v; end`, rather than copying the whole structure followed by rehashing it.

**Review by JeffBezanson — APPROVED** [review]

**Comment by bkamins** [comment]
Good idea - I will do some benchmarks to find a reasonable break even and propose another PR.

**Comment by bkamins** [comment]
CI failure seems unrelated. Should this be merged?

**Pull request merged**